app/models/date_validation.rb

```diff
@@ -0,0 +1,62 @@
+module DateValidation
+  extend ActiveSupport::Concern
+
+  attr_reader :invalid_date_attributes
+
+  included do
+    validates_with DateValidator
+
+    after_validation :rationalise_date_errors
+
+  private
+
+    # In cases when a date attribute is invalid, it will be set to nil by pre_validate_date_attribute and will therefore
+    # fail the presence validation. We therefore need to remove the presence error from each invalid date attribute to
+    # avoid a confusing user experience where both the invalid date and presence errors show simultaneously
+    def rationalise_date_errors
+      return if invalid_date_attributes.nil?
+
+      invalid_date_attributes.each do |invalid_date_attribute|
+        if errors.of_kind?(invalid_date_attribute, :blank)
+          errors.delete(invalid_date_attribute, :blank)
+        end
+      end
+    end
+  end
+
+  def pre_validate_date_attribute(attribute, date)
+    if date.is_a?(Hash)
+      begin
+        raise ArgumentError if date.values.any?(&:nil?)
+        raise ArgumentError if date[1].to_i.zero? || date[2].to_i.zero? || date[3].to_i.zero?
```

Comment:
This might be more succinct:

```suggestion
        raise ArgumentError if date.values.any? { |v| v.to_i.zero? }
```

But just so I understand: what scenario is this protecting against?

These values should already have been cast to integers at this point, which is why `.to_i` hasn't been called on line 33 (i.e. the `Date.new` line). Non-numeric values will have become `0`.

The `Date.new` constructor already rejects `0` input values:

```ruby
Date.new(2023, 1, 0)
# => Date::Error: invalid date
```

So I'm not sure if this guard clause protects against a scenario that wasn't already covered? 🤔

Reply:
I used a loop originally, then I had to do this: https://github.com/alphagov/whitehall/pull/8221

Reply:
I've found a better way to do this by catching TypeErrors which are triggered when a non-integer value is passed to the date constructor. Update to follow